xdem/coreg.py

```diff
@@ -1712,3 +1714,144 @@ def _to_matrix_func(self) -> np.ndarray:
         matrix[2, 3] += self._meta["bias"]
 
         return matrix
+
+
+def invert_matrix(matrix: np.ndarray) -> np.ndarray:
+    """Invert a transformation matrix."""
+    with warnings.catch_warnings():
+        # Deprecation warning from pytransform3d. Let's hope that is fixed in the near future.
+        warnings.filterwarnings("ignore", message="`np.float` is a deprecated alias for the builtin `float`")
+
+        # Invert the transform if wanted.
+        return pytransform3d.transformations.invert_transform(matrix)
+
+
+def _create_nan_mask(array: Union[np.ndarray, np.ma.masked_array]) -> np.ndarray:
+    """Will be replaced by Amaury's version once that is merged."""
+    nans = ~np.isfinite(np.asarray(array))
+
+    if isinstance(array, np.ma.masked_array):
+        nans = nans | array.mask
+
+    return nans
+
+
+def apply_matrix(dem: np.ndarray, transform: rio.transform.Affine, matrix: np.ndarray, invert: bool = False,
```

Comment:
This is great work, yet I keep thinking there must be a way to apply all of this simultaneously (and that would probably be more efficient computationally). There's nothing like this in pytransform3d?

Reply:
Thanks! I am certain there may be a simpler way mathematically:

"3D rigid transform"  -> "2D projective transform" + "pixel value correction"

I'm just not that good at math, unfortunately!